fud2/fud-core/src/run.rs

```diff
@@ -87,6 +87,77 @@ impl EmitBuild for EmitBuildFn {
     }
 }
 
+/// A config variable.
+#[derive(Debug, Clone)]
+pub enum ConfigVar {
+    /// The key for the config variable.
+    Var(String),
+    /// The key for the config variable followed by the value it should be if the key is not found.
+    VarOr(String, String),
+}
+
+/// The data required to emit a single op.
+pub struct OpEmitData {
+    /// The name of the op.
+    pub op_name: String,
+
+    /// The commands run whenever the op run. Each of these must be run in order in a context where
+    /// the variables in each cmd are supplied. In particular, this means that variables of the
+    /// form "$[i|o]<digit>" are in scope.
+    pub cmds: Vec<String>,
+
+    /// Variables to be emitted
+    pub config_vars: Vec<ConfigVar>,
+}
+
+/// Given the `index` of a file in the list of input/output files and if the file is an
+/// `input`, returns a valid Ninja variable name.
+pub fn io_file_var_name(index: usize, input: bool) -> String {
+    if input {
+        format!("i{}", index)
+    } else {
+        format!("o{}", index)
+    }
+}
+
+impl EmitBuild for OpEmitData {
+    fn build(
+        &self,
+        emitter: &mut StreamEmitter,
+        inputs: &[&str],
+        outputs: &[&str],
+    ) -> EmitResult {
+        // Write the Ninja file.
+        let cmd = self.cmds.join(" && ");
+        emitter.rule(&self.op_name, &cmd)?;
+        let in_vars = inputs
+            .iter()
+            .enumerate()
+            .map(|(k, &v)| (io_file_var_name(k, true), v));
+        let out_vars = outputs
+            .iter()
+            .enumerate()
+            .map(|(k, &v)| (io_file_var_name(k, false), v));
+        let vars: Vec<_> = in_vars.chain(out_vars).collect();
```

Comment:
I suppose that we wouldn't need the `collect` here if we just made a bunch of calls to `emitter.arg(k, v)` for each of these, after emitting the `build_cmd`…

Reply:
Oh, that is true. I'm going to keep this as is though because I can see the future (i.e. a currently open pull request) and know making `vars` and then calling `build_cmd_with_args` is probably better when multiple build statements are needed per op.

Reply:
Sounds good. Always something that we could consider "streamifying" later, but it's not worth the churn for now.